node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/ZoneAppMigrationTest.java

```diff
@@ -0,0 +1,171 @@
+        assertEquals(20, tester.nodeRepository().getNodes(NodeType.host, Node.State.active).size());
+        assertEquals(tenantNodes, Set.copyOf(tester.nodeRepository().getNodes(NodeType.tenant)));
+
+        // All tenant hosts are allocated to tenant host application
+        assertEquals(Set.copyOf(tester.nodeRepository().getNodes(NodeType.host)),
+                Set.copyOf(tester.nodeRepository().getNodes(tenantHostApp)));
+
+        // All proxy nodes are still allocated to zone-app
+        assertEquals(Set.copyOf(tester.nodeRepository().getNodes(NodeType.proxy)),
+                Set.copyOf(tester.nodeRepository().getNodes(zoneApp)));
+    }
+
+    @Test
+    public void conflicting_type_allocation_test() {
+        // Re-allocate tenant host from zone-app to tenant-host app
+        tester.activate(zoneApp, prepareSystemApplication(zoneApp, NodeType.proxy, "routing"));
+        inactiveExpirer.maintain();
+        tester.nodeRepository().getNodes(NodeType.host)
+                .forEach(node -> tester.nodeRepository().setReady(node.hostname(), Agent.operator, "Readied by host-admin"));
+        tester.activate(tenantHostApp, prepareSystemApplication(tenantHostApp, NodeType.host, "tenant-host"));
+
+        // Re-deploying zone-app with both type proxy and host has no effect (no tenant hosts are re-allocated from tenant-host app)
+        Set<Node> allNodes = Set.copyOf(tester.nodeRepository().getNodes());
+        List<HostSpec> proxyHostSpecs = prepareSystemApplication(zoneApp, NodeType.proxy, "routing");
+        List<HostSpec> nodeAdminHostSpecs = prepareSystemApplication(zoneApp, NodeType.host, "node-admin");
+        List<HostSpec> zoneAppHostSpecs = concat(proxyHostSpecs, nodeAdminHostSpecs, Collectors.toList());
+        tester.activate(zoneApp, zoneAppHostSpecs);
+        assertEquals(0, nodeAdminHostSpecs.size());
+        assertEquals(allNodes, Set.copyOf(tester.nodeRepository().getNodes()));
+
+        // Provision another host and redeploy zone-app
+        Node newHost = tester.makeReadyNodes(1, "large", NodeType.host).get(0);
+        proxyHostSpecs = prepareSystemApplication(zoneApp, NodeType.proxy, "routing");
+        nodeAdminHostSpecs = prepareSystemApplication(zoneApp, NodeType.host, "node-admin");
+        zoneAppHostSpecs = concat(proxyHostSpecs, nodeAdminHostSpecs, Collectors.toList());
+        tester.activate(zoneApp, zoneAppHostSpecs);
+
+        assertEquals(1, nodeAdminHostSpecs.size()); // The newly provisioned host is prepared
+        newHost = tester.nodeRepository().getNode(newHost.hostname()).orElseThrow(); // Update newHost after it has been allocated
+        Set<Node> allNodesWithNewHost = concat(allNodes, Set.of(newHost), Collectors.toSet());
+        assertEquals(allNodesWithNewHost, Set.copyOf(tester.nodeRepository().getNodes()));
+        // The new host is allocated to zone-app, while the old ones are still allocated to tenant-host app
```

Comment:
And these will be inactivated, dirtied, readied, and activated into the tenant-host at the next release w/o the type=host cluster?

Reply:
Yes